engine/src/main/java/org/hibernate/validator/spi/cascading/ValueExtractor.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.spi.cascading;
+
+public interface ValueExtractor<T> {
+
+	void extractValues(T originalValue, ValueReceiver receiver);
+
+	interface ValueReceiver {
+
+		void objectValue(Object object);
+
+		void listValue(int i, Object object);
```

Comment:
Yes, "listItem" would be better.

That said, the entire contract needs to change anyways, it needs to be more generic. That's one of the next things to look into. But essentially we don't know the "nature" of the association navigated into, so we cannot work with a fixed set of pre-defined methods. One idea is to the creation of `Node` instances to that contract, but we'll have to see.